### PR TITLE
New plan for p2g

### PR DESCRIPTION
**Pull request by zeshunzong**
This change is [Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)

### CODE REVIEW

**Review by zeshunzong — COMMENTED** [review]
+@xuchenhan-tri Take a look at the finished Particles class. Tests are also updated. I have to create a new PR as I accidentally at pre which adds the guy named pre into our discussion.

I will have P2G finished ASAP.

> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-NjLssRp3fpRN4Ele-PS:b10tdxu)* status: 0 of 8 files reviewed, all discussions resolved (waiting on @xuchenhan-tri)

**Review by xuchen-han — CHANGES_REQUESTED** [review]
I reviewed the Particles class. It looks fine. I have a few minor comments.
Please PR only relevant files.  For example, this PR should probably only contain `particles.h/cc` and its test.
It helps the reviewer as well as yourself to reason about the design of individual components.

> Reviewed 4 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-NjOe9yO4VEF40oXEvhx:b1k7un0)* status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @zeshunzong)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOdYL75khBzkry7l5f:-NjOdYL75khBzkry7l5g:b1zdwc3) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/interpolation_weights.h#L44)):*
> ```objc
> 
>   /**
>    * Empty constructor.
> ```

Documentations like this are useless. Any reader of this code would understand that this is the default constructor. The documentation should focus on the implication of a default constructed `InterpolationWeights`. In particular, the weights are invalid, so why does this constructor exist?
___
*[`multibody/mpm/particles.h` line 65 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObPfG3piTpuxopXCX:-NjObPfG3piTpuxopXCY:b-gj3m5v) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L65)):*
> ```objc
>    * Finalize(); it serves more like an internal checker.
>    */
>   void Finalize();
> ```

I'm not seeing why this function is needed. It doesn't finalize anything. Can you explain?
___
*[`multibody/mpm/particles.h` line 73 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObZWD4IxR19mNek-C:-NjObZWD4IxR19mNek-D:b-imzifz) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L73)):*
> ```objc
>    * 3) Computes batch_starts_ and batch_sizes_.
>    * 4) Compute weights and weight gradients
>    */
> ```

Right now the documentation is an implementation notes which could be useful. However, it's very light on why this is needed and how it should be used. What kind of invariants can a user expect after calling this?



_Code quote:_
```objc
  /**
   * 1) Computes the base node for each particle.
   * 2) Sorts particle attributes lexicographically by their base nodes
   * positions.
   * 3) Computes batch_starts_ and batch_sizes_.
   * 4) Compute weights and weight gradients
   */
```
___
*[`multibody/mpm/particles.h` line 106 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc20QBkjJFgG9i_Vw:-NjOc20QBkjJFgG9i_Vx:b-evug5k) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L106)):*
> ```objc
>    * 
>    * 
> ```

nit why not make it private already?
___
*[`multibody/mpm/particles.cc` line 56 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOcokK9odG4T6CwOA4:-NjOcokK9odG4T6CwOA5:b-f3ceta) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.cc#L56)):*
> ```cpp
> void Particles<T>::Prepare(double h) {
>   DRAKE_DEMAND(num_particles() > 0);
>   // 1) compute the base node for each particle
> ```

What's the purpose of these numbers (1, 2, 2.1, 2.2, etc.)? It'd be nice if they map 1-1 to the documented responsibility of this function in the header.
___
*[`multibody/mpm/particles.cc` line 145 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc_2d0IDlrAVHCtuU:-NjOc_2d0IDlrAVHCtuV:b8zgp5e) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.cc#L145)):*
> ```cpp
>   DRAKE_DEMAND((new_order.size()) == num_particles());
> 
>   for (size_t i = 0; i < num_particles(); ++i) {
> ```

BTW, the amount of repetition here is screaming for a data managing class.
___
*[`multibody/mpm/internal/math_utils.h` line 140 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOd8mh-ztNUs4pT8Rs:-NjOd8mh-ztNUs4pT8Rt:b52ai23) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/internal/math_utils.h#L140)):*
> ```objc
> }
> 
> // Returns true if index_1 < index_2
> ```

I feel like I have reviewed these changes before. Why are they showing up again as new diffs?

**Review by zeshunzong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)* status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @xuchenhan-tri)

___
*[`multibody/mpm/particles.h` line 65 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObPfG3piTpuxopXCX:-NjOg0mI4yC7SzTjiQXk:b-f77gqp) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L65)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I'm not seeing why this function is needed. It doesn't finalize anything. Can you explain?
</blockquote></details>

My design is that this is called after all AddParticles() are done. Two things I want to do, first check all dimensions are correct (for those attributes that have size num_particles()). 

Second, I want to reserve space for the batch-based attributes. They can be upper-bounded by the number of particles.
___
*[`multibody/mpm/particles.h` line 106 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc20QBkjJFgG9i_Vw:-NjOfEOV-DF3Cs6qhA1i:b-c7icl7) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L106)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

nit why not make it private already?
</blockquote></details>

There are tests wrote for the Reorder functions. (which is not necessarily needed from a global point of view, as the larger Prepare() function has been tested) I decided to keep the tests. Once we made a final decision on the choice of Reorder function, we can remove the tests and make them private. I just don't bother change the soon-to-be-deleted test functions.

**Review by xuchen-han — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)* status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @zeshunzong)

___
*[`multibody/mpm/particles.h` line 65 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObPfG3piTpuxopXCX:-NjOhMt3-cxzvC9rX41I:bj5nhsu) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L65)):*
<details><summary><i>Previously, zeshunzong (Zeshun Zong) wrote…</i></summary><blockquote>

My design is that this is called after all AddParticles() are done. Two things I want to do, first check all dimensions are correct (for those attributes that have size num_particles()). 

Second, I want to reserve space for the batch-based attributes. They can be upper-bounded by the number of particles.
</blockquote></details>

1. The consistency of the dimension should be maintained as a class invariant. If they are not consistent, you have a bug in your code. if the only purpose of the function is invariant checking, it shouldn't be a user-callable public function and it should be done automatically.

2. Re reserving spaces: you can reserve the memory right when you use it. No need for a special Finalize() function to do it.

TLDR, the Finalize() is badly named and shouldn't be part of you public API.
___
*[`multibody/mpm/particles.h` line 106 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc20QBkjJFgG9i_Vw:-NjOibO3EtqDb5RKMi3D:b-9alrpo) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L106)):*
<details><summary><i>Previously, zeshunzong (Zeshun Zong) wrote…</i></summary><blockquote>

There are tests wrote for the Reorder functions. (which is not necessarily needed from a global point of view, as the larger Prepare() function has been tested) I decided to keep the tests. Once we made a final decision on the choice of Reorder function, we can remove the tests and make them private. I just don't bother change the soon-to-be-deleted test functions.
</blockquote></details>

"I need to unit test the function" is not the reason to expose internal logic to public API. You can friend the testing class for that.

**Review by zeshunzong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)* status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @xuchenhan-tri)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOdYL75khBzkry7l5f:-NjTbWqe9R8nvdxcTtNV:b-nei1rr) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/interpolation_weights.h#L44)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Documentations like this are useless. Any reader of this code would understand that this is the default constructor. The documentation should focus on the implication of a default constructed `InterpolationWeights`. In particular, the weights are invalid, so why does this constructor exist?
</blockquote></details>

I do agree that the documentation is not necessary. I disagree with the statement that since weights are invalid, they cannot be default constructed. Think of them as placeholders. There are different ways that weights can be invalid. I don't think they need to be constructed as valid numbers at the very beginning. 

I instead offer a boolean is_valid_ to guard their validity. See the newly added change.
___
*[`multibody/mpm/particles.h` line 65 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObPfG3piTpuxopXCX:-NjTUhz12c_i92HZ6s07:behd7n4) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L65)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

1. The consistency of the dimension should be maintained as a class invariant. If they are not consistent, you have a bug in your code. if the only purpose of the function is invariant checking, it shouldn't be a user-callable public function and it should be done automatically.

2. Re reserving spaces: you can reserve the memory right when you use it. No need for a special Finalize() function to do it.

TLDR, the Finalize() is badly named and shouldn't be part of you public API.
</blockquote></details>

removed. the checker is relocated to AddParticles() --- which I agree is unnecessary, I add a todo to indicate this is temporary safeguard during development. Reservation is put at the front of Prepare() (which I think will be called only once in its first function call, check that)
___
*[`multibody/mpm/particles.h` line 73 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjObZWD4IxR19mNek-C:-NjTbUfu5Go6WZZt2doz:bb7f2mf) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L73)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Right now the documentation is an implementation notes which could be useful. However, it's very light on why this is needed and how it should be used. What kind of invariants can a user expect after calling this?
</blockquote></details>

done
___
*[`multibody/mpm/particles.h` line 106 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc20QBkjJFgG9i_Vw:-NjTeL062d2jIodBoT5w:b8sn9oo) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.h#L106)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

"I need to unit test the function" is not the reason to expose internal logic to public API. You can friend the testing class for that.
</blockquote></details>

I am aware of these. It is just that I know those (local) tests will be removed and a friend class is eventually not necessary. I feel like I might eventually forget to take off the friend signature and this looks ugly to me.

I offer the alternative. I removed the customized tests for Reorder() and Reorder2(), and we can add it back when we study the reorder algorithms in details later. 

I made them private. Meanwhile, they are still being tested and validated as the function that utilizes the reordering is tested.
___
*[`multibody/mpm/particles.cc` line 56 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOcokK9odG4T6CwOA4:-NjT__o6B2pnVQJlUA8S:blqgb34) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.cc#L56)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

What's the purpose of these numbers (1, 2, 2.1, 2.2, etc.)? It'd be nice if they map 1-1 to the documented responsibility of this function in the header.
</blockquote></details>

done they do now
___
*[`multibody/mpm/particles.cc` line 145 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOc_2d0IDlrAVHCtuU:-NjTVaztEiS5vhCycXtw:b7ghn4i) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/particles.cc#L145)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, the amount of repetition here is screaming for a data managing class.
</blockquote></details>

I know. For the issue itself, I would postpone it later. On the other hand, I am more and more in favor of array of struct (each particle) now.
___
*[`multibody/mpm/internal/math_utils.h` line 140 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOd8mh-ztNUs4pT8Rs:-NjTVxHi7ZZEKoIn4GuC:b-4m3f8y) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/internal/math_utils.h#L140)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I feel like I have reviewed these changes before. Why are they showing up again as new diffs?
</blockquote></details>

minor they are not merged to the main branch. And I start afresh with a new branch.

**Review by xuchen-han — CHANGES_REQUESTED** [review]
> Reviewed 2 of 8 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)* status: all files reviewed, 2 unresolved discussions (waiting on @zeshunzong)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOdYL75khBzkry7l5f:-NjiqwMW57h_DfsjIN8M:bo8274t) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/interpolation_weights.h#L44)):*
> Think of them as placeholders. 

I don't understand why you need these placeholders to begin with. Each flag you are adding to your system (like `is_valid` or `needs_reordering`) is making the code more brittle. They are essentially manually managed cache entries. To maintain your code, we now need to remember when and where each cache needs to be invalidated. 

I'm guessing the reason that you need a placeholder is that you don't have the means to compute the weights at initialization? If so, why not just leaving `Particles::weights_` empty?
___
*[`multibody/mpm/particles.cc` line 51 at r2](https://reviewable.io/reviews/zeshunzong/drake/21#-NjisbBu7u5p71UGyN7Y:-NjisbBvDAvt5RGNU5iQ:b-5xunno) ([raw file](https://github.com/zeshunzong/drake/blob/3036bbf50103b05a74d2fb592a2106f5e6d1a1ec/multibody/mpm/particles.cc#L51)):*
> ```cpp
>   // reserve space for batch_starts_ and batch_sizes
>   // they can be upper-bounded by num_particles() as there is no empty batch
>   if (batch_starts_.size() == 0) {
> ```

minor, I don't think this predicate is necessary. You can safely call reserve at all time steps.
___
*[`multibody/mpm/internal/math_utils.h` line 140 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOd8mh-ztNUs4pT8Rs:-NjiozZy74iksRbfo4Ed:bbgqaky) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/internal/math_utils.h#L140)):*
<details><summary><i>Previously, zeshunzong (Zeshun Zong) wrote…</i></summary><blockquote>

minor they are not merged to the main branch. And I start afresh with a new branch.
</blockquote></details>

This is another reason for smaller PRs: they get merged a lot more easily.

**Review by zeshunzong — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-Njj1GHs1Sfh5F0SwMdE:bb7f2mf)* status: 4 of 8 files reviewed, 2 unresolved discussions (waiting on @xuchenhan-tri)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOdYL75khBzkry7l5f:-Njj0fNe-Y1X2FpNY0Ui:bcz2gr3) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/interpolation_weights.h#L44)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

> Think of them as placeholders. 

I don't understand why you need these placeholders to begin with. Each flag you are adding to your system (like `is_valid` or `needs_reordering`) is making the code more brittle. They are essentially manually managed cache entries. To maintain your code, we now need to remember when and where each cache needs to be invalidated. 

I'm guessing the reason that you need a placeholder is that you don't have the means to compute the weights at initialization? If so, why not just leaving `Particles::weights_` empty?
</blockquote></details>

done boolean removed
___
*[`multibody/mpm/particles.cc` line 51 at r2](https://reviewable.io/reviews/zeshunzong/drake/21#-NjisbBu7u5p71UGyN7Y:-Njj0lpw61-_ESBj8t1A:bhaj9rj) ([raw file](https://github.com/zeshunzong/drake/blob/3036bbf50103b05a74d2fb592a2106f5e6d1a1ec/multibody/mpm/particles.cc#L51)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

minor, I don't think this predicate is necessary. You can safely call reserve at all time steps.
</blockquote></details>

done removed. question: any performance difference between the two versions?
___
*[`multibody/mpm/particles.cc` line 90 at r3](https://reviewable.io/reviews/zeshunzong/drake/21#-Njj0tQO0-_JfAe7l7Mp:-Njj0tQO0-_JfAe7l7Mq:b7fcpm9) ([raw file](https://github.com/zeshunzong/drake/blob/c499ba578441902dd72895ab5213c7552d53fbcd/multibody/mpm/particles.cc#L90)):*
> ```cpp
> 
>   // 4) compute d and dw
>   for (size_t p = 0; p < num_particles(); ++p) {
> ```

typo, spotted when I test for p2g
___
*[`multibody/mpm/internal/math_utils.h` line 140 at r1](https://reviewable.io/reviews/zeshunzong/drake/21#-NjOd8mh-ztNUs4pT8Rs:-Njj10JP88ySTQAxyswq:bo23su9) ([raw file](https://github.com/zeshunzong/drake/blob/ff32eedc30953c1097fab98d68283920049e2398/multibody/mpm/internal/math_utils.h#L140)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

This is another reason for smaller PRs: they get merged a lot more easily.
</blockquote></details>

done acknowledged

**Review by xuchen-han — CHANGES_REQUESTED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-Njj4m2n44aDsvSzBxXf:bg7igkv)* status: all files reviewed, 1 unresolved discussion (waiting on @zeshunzong)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r3](https://reviewable.io/reviews/zeshunzong/drake/21#-Njj4ga5D4ie-cOHIq43:-Njj4ga6-Ke6U9yv-C6M:bb7phvt) ([raw file](https://github.com/zeshunzong/drake/blob/c499ba578441902dd72895ab5213c7552d53fbcd/multibody/mpm/interpolation_weights.h#L44)):*
> ```objc
>   /**
>    * All weights and weight gradients are initialized to be zero. The weights
>    * will be valid until Reset() is called.
> ```

typo



_Suggestion:_
```objc
  /**
   * All weights and weight gradients are initialized to be zero. The weights
   * will be invalid until Reset() is called.
```
___
*[`multibody/mpm/particles.cc` line 51 at r2](https://reviewable.io/reviews/zeshunzong/drake/21#-NjisbBu7u5p71UGyN7Y:-Njj44xi-mTMIokYcI9l:b-7jvzvv) ([raw file](https://github.com/zeshunzong/drake/blob/3036bbf50103b05a74d2fb592a2106f5e6d1a1ec/multibody/mpm/particles.cc#L51)):*
<details><summary><i>Previously, zeshunzong (Zeshun Zong) wrote…</i></summary><blockquote>

done removed. question: any performance difference between the two versions?
</blockquote></details>

There's none. It's a no-op most of the time: https://cplusplus.com/reference/vector/vector/reserve/
___
*[`multibody/mpm/particles.cc` line 90 at r3](https://reviewable.io/reviews/zeshunzong/drake/21#-Njj0tQO0-_JfAe7l7Mp:-Njj4YzZDu6i2cXZPs3A:b6zrrk4) ([raw file](https://github.com/zeshunzong/drake/blob/c499ba578441902dd72895ab5213c7552d53fbcd/multibody/mpm/particles.cc#L90)):*
<details><summary><i>Previously, zeshunzong (Zeshun Zong) wrote…</i></summary><blockquote>

typo, spotted when I test for p2g
</blockquote></details>

Good catch!

**Review by xuchen-han — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-Njj4oMX1393IJOAItOI:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @zeshunzong)

**Review by zeshunzong — COMMENTED** [review]
typo fixed, request merge

> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21#-:-Njj66-z7aHY6-zIq5su:b-qo24y2)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @zeshunzong)

___
*[`multibody/mpm/interpolation_weights.h` line 44 at r3](https://reviewable.io/reviews/zeshunzong/drake/21#-Njj4ga5D4ie-cOHIq43:-Njj5sjq3zIN9HMhFyOG:bb7f2mf) ([raw file](https://github.com/zeshunzong/drake/blob/c499ba578441902dd72895ab5213c7552d53fbcd/multibody/mpm/interpolation_weights.h#L44)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

typo
</blockquote></details>

done
___
*[`multibody/mpm/particles.cc` line 51 at r2](https://reviewable.io/reviews/zeshunzong/drake/21#-NjisbBu7u5p71UGyN7Y:-Njj5zFM2NGEyKwohAR9:bb7f2mf) ([raw file](https://github.com/zeshunzong/drake/blob/3036bbf50103b05a74d2fb592a2106f5e6d1a1ec/multibody/mpm/particles.cc#L51)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

There's none. It's a no-op most of the time: https://cplusplus.com/reference/vector/vector/reserve/
</blockquote></details>

done
___
*[`multibody/mpm/particles.cc` line 90 at r3](https://reviewable.io/reviews/zeshunzong/drake/21#-Njj0tQO0-_JfAe7l7Mp:-Njj5zyu-ayjx6vFcAT7:bb7f2mf) ([raw file](https://github.com/zeshunzong/drake/blob/c499ba578441902dd72895ab5213c7552d53fbcd/multibody/mpm/particles.cc#L90)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Good catch!
</blockquote></details>

done

**Review by xuchen-han — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeshunzong/drake/21)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @zeshunzong)
